rir/src/interpreter/interp.cpp

```diff
@@ -305,9 +305,19 @@ SEXP createLegacyArgsList(const CallContext& call, bool eagerCallee,
     return result;
 }
 
-SEXP createLegacyLazyArgsList(const CallContext& call, Context* ctx) {
+SEXP argsLazyCreation(void* rirDataWrapper) {
+    ArgsLazyData* argsLazy = ArgsLazyData::unpack(rirDataWrapper);
+    return argsLazy->createArgsLists();
+}
+
+SEXP createLegacyLazyArgsList(const CallContext& call, Context* ctx,
+                              bool needsEnv) {
     if (call.hasStackArgs()) {
-        return createLegacyArgsListFromStackValues(call, false, ctx);
+        if (needsEnv)
+            return createLegacyArgsListFromStackValues(call, false, ctx);
+        else
+            // double case to avoid alignment warning
+            return (SEXP)(void*)(new ArgsLazyData(&call, ctx));
```

Comment:
Oh, this was totally on purpose. I know about `RirRuntimeObjects`. I created `ArgsLazyData` based on them: https://github.com/charig/rir/blob/lazyArgs/rir/src/runtime/ArgsLazyData.h.

The main difference is that they are not wrapped inside a `SEXP`. My point here is that the main goal of this PR is to avoid creating an unnecessary `SEXP` on every function call for the arguments. If we avoid that but need to create another SEXP for wrapping the data that is needed to lazily create the SEXP, and need to add some checks here and there to test what kind of argsList we really created, then what do we gain?
I took some hours to read the gnu-r code and it seemed that the RCNTXT are managed in ad-hoc manners (even by the gc), i.e, https://github.com/reactorlabs/gnur/blob/lazyPromargs/src/main/memory.c#L1693.

I resort on the gnu-r test suite that triggers some GCs for testing. Anyway, if this is not safe it would be very easy for me to make `ArgsLazyData` to subclassify `RirRuntimeObjects` instead of `RirDataWrapper`. But then I guess the PR lose much of its value.

Reply:
> Ok I see. Though, it would still help, since it reduces the allocations from the whole argslist to just one cell.
> 
> Fair enough, I am conviced that we can get away with not allocating an R object, but just an internal marker thingy.
> 
> Some additional comments though:
> 
> * [this](https://github.com/reactorlabs/rir/pull/303/files#diff-48bddc5b82508439eaadd23792d2851bR664) is wrong, since it might be replaced by an actual SEXP during the call. Why not use stack allocated memory in the first place? In doCall just do:
> 
> ```
> ArgsLazyData lazyArgRecreateData(call, ctx);
> arglist = &lazyArgRecreateData;
> ```

It is not. argsList always points to the `ArgsLazy` wrapper. The field that is actually replaced in place in case we need the promargs is the RCNTXT->promargs. I can not do what you suggest in `doCall` because at that point I do not know what kind of argsList I would need. To avoid manual memory and leverage on the stack `rirCall` needs a major refactoring to abstract the argList creation for anycase. 
  

> * [this](https://github.com/reactorlabs/rir/pull/303/files#diff-48bddc5b82508439eaadd23792d2851bR315) is not a good idea. Imho the warning is real. You should probably remove the pragma pack stuff instead.
> * [this](https://github.com/reactorlabs/gnur/blob/lazyPromargs/src/main/context.c#L121) looks brittle. Could you at least put the marker at the beginning of the struct? And then on the rir side you could add a static assert. Something like:
> 
> ```
> RirDataWrapper(uint32_t count) : info{count, MAGIC} {
>     //needs to be in sync with gnur context.c
>    static_assert(&this->magic == &((utint32_t*)this));
> }
> ```

Reply:
> It is not. argsList always points to the ArgsLazy wrapper.

my bad, sorry, you are right

> I can not do what you suggest in doCall

sure, just put `ArgsLazyData lazyArgRecreateData(call, ctx);` below line 655 and only use it in one branch

Reply:
(btw. the problem with new/free is also, that we will leak when a non-local return jumps over the delete)